StackLink and getTargetUrl

thomasdax98 opened the pull request:
Draft for `StackLink` and `getTargetUrl`

- `StackLink` provides a react-router link wrapped in a Mui Link for navigating in a stack
- `stackApi.getTargetUrl` returns the url of a stack page that can be navigated to
- `StackLink` and `getTargetUrl` have the same parameters as `stackApi.activatePage`

What do you think about that?

Todo:
- [x] Changelog
- [x] Docs

thomasdax98 commented on packages/admin/src/stack/Switch.tsx:
```diff
@@ -49,6 +53,9 @@ export function useStackSwitch(): [React.ComponentType<IProps>, IStackSwitchApi]
             apiRef.current?.activatePage(pageName, payload, subUrl);
         },
+        getTargetUrl: (pageName: string, payload: string, subUrl?: string) => {
+            return apiRef.current?.getTargetUrl(pageName, payload, subUrl) || "";
```
Is this a good way of handling the "undefined"-case or should I rather throw an exception?

nsams replied:
good question. Or return "" and log an console error.

nsams commented:
lgtm

nsams commented:
And i'm currently writing on Stack docs, you can integrate the new component in there when it's ready.

nsams commented:
https://github.com/vivid-planet/comet-admin/pull/446 is merged, you can add your docs in there now

thomasdax98 commented:
@nsams @rabengraph I slightly changed the StackLink and added a prop for passing an external switchApi. That way it's also possible to use the StackLink outside of a stack (e.g. in combination with the useStackSwitch hook)

Also, I added docs for StackLink and getTargetUrl. And I added props tables to the docs.

nsams commented on packages/admin-stories/src/docs/components/Stack/Stack.stories.mdx:
```diff
@@ -49,13 +79,48 @@ This method will activate (navigate to) the given page.
+| :--------- | :----- | :------------ | :------------------------------------------------------------------------------- |
+| pageName\* | string |               | Name of the page that will be activated.                                         |
+| payload\*  | string |               | Additional payload containing eg. an id that will get passed into the StackPage. |
+| subUrl     | string |               | SubUrl to append, for nested stacks or other nested routes.                      |
```
I would remove the Default value column if there is no single default value in the table

nsams commented on packages/admin-stories/src/docs/components/Stack/Stack.stories.mdx:
```diff
@@ -49,13 +79,48 @@ This method will activate (navigate to) the given page.
+| subUrl     | string          |               | SubUrl to append, for nested stacks or other nested routes.                                    |
+| switchApi  | IStackSwitchApi |               | Custom switchApi, has to be passed if the StackLink is used outside of a StackSwitch component |
+
+### getTargetUrl
```
is that method worth documenting? (In pratice you will always use a StackLink)

(it could be mentioned as a side-note tough)

thomasdax98 replied:
I removed the section. Instead I mention it in the activatePage section now

nsams commented on packages/admin-stories/src/docs/components/Stack/Stack.stories.tsx:
```diff
@@ -280,4 +288,24 @@ storiesOf("stories/components/Stack", module)
             </div>
         );
+    })
+    .add("StackLink", () => {
```
I would even say we should promote StackLink as the default solution, also meaning all existing examples should be ported to it. (but that can happen in it's own PR)

thomasdax98 replied:
Okay, I'll create an extra PR for that

nsams commented on packages/admin/src/stack/StackLink.tsx:
```diff
@@ -0,0 +1,30 @@
+    children,
+}: PropsWithChildren<StackLinkProps>): React.ReactElement => {
+    const internalSwitchApi = useStackSwitchApi();
+    // external switchApi allows the creation of StackLinks outside of the stack with the useStackSwitch() hook
```
do you have a real use case for that?

thomasdax98 replied:
I tried to rebuild the examples in the stories with the StackLink. Here I couldn't do it, because the button is outside StackSwitch:

```tsx
// Original
const [StackSwitch, switchApi] = useStackSwitch();
return (
    <div>
        <button
            onClick={() => {
                switchApi.activatePage("page2", "foo");
            }}
        >
            page2
        </button>
        <Stack topLevelTitle="Stack">
            <StackBreadcrumbs />
            <StackSwitch>
                <StackPage name="page1">
                    <div>page1</div>
                </StackPage>
                <StackPage name="page2">
                    <div>page2</div>
                </StackPage>
            </StackSwitch>
        </Stack>
    </div>
);
```

```tsx
// With StackLink
const [StackSwitch, switchApi] = useStackSwitch();
return (
    <div>
        {/*This StackLink has is outside StackSwitch => has no StackSwitchContext => doesn't work*/}
        <StackLink pageName="page2" payload="foo">
             link to page2
        </StackLink>
        <Stack topLevelTitle="Stack">
            <StackBreadcrumbs />
            <StackSwitch>
                <StackPage name="page1">
                    <div>page1</div>
                </StackPage>
                <StackPage name="page2">
                    <div>page2</div>
                </StackPage>
            </StackSwitch>
        </Stack>
    </div>
);
```

Therefore I added the `switchApi` prop, so that it's possible to combine the `StackLink` with `useStackSwitch`:

```
<StackLink pageName="page2" payload="foo" switchApi={switchApi}>
    link to page2
</StackLink>
```

thomasdax98 replied:
I haven't yet looked if there is a specific case in Comet where this is needed. I can remove it for now and make an extra PR if it's needed

nsams replied:
yes, makes sense.

I also don't know if there is a real use case for that, but on the other side it won't do any harm or make the api ugly
